ssd: added documentation on enabling trim

anjandev opened the pull request.

flexibeast commented:
Looks like you didn't run your work through `vmdfmt` before committing, as per [the style guide](https://docs.voidlinux.org/contributing/void-docs/style-guide.html)?

anjandev commented:
@flexibeast my apologies. I fixed it with the latest force push.

flexibeast commented:
Thanks. :-)

Some initial comments:

* Please make sure the text consistently refers to 'TRIM' rather than 'trim', unless it's in the context of something to be typed on the command line.
* Line 23: Change this to remove the pronoun, and rewrite it in the imperative: "To automate running TRIM, use \[cron\]\(../config/cron.md\), or add the \`discard\` option to \`/etc/fstab\`."
* Line 27: Remove "To automate trim, we can use \[cron\]\(../config/cron.md\)."
* Line 36: Remove "by running the following command".
* Line 44: Capitalise 'Linux'.
* Line 50: Remove "I would" and "as such".
* Line 96: "ie." should probably be "e.g.".

anjandev commented:
@flexibeast thank you for the feedback. All your suggestions are great and I implemented them in the latest commit. Would you mind using github's review functionality to comment so that it's easier to find the line you are referencing?

flexibeast commented:
@anjandev: My apologies! Sure, i can do that in the future. :-) However, i think i'm now at the limit of the feedback i can provide, as i'm not a member of the Void team.

Thanks for your contribution!

anjandev commented:
@flexibeast Ya, I might add some more docs. I spent like a month trying to figure out why programs and my computer was randomly crashing until I realized that the void handbook doesnt remind you to enable trim. It was kind of frustrating and relieving at the same time.

q66 commented:
Would be good to add instructions for enabling TRIM in LVM/LUKS setups. Also, the comment about using a FS that supports TRIM is probably useless, all commonly used filesystems do, it's mostly not a filesystem thing.

anjandev commented:
@q66 I added documentation for LVM/LUKS and ZFS. Im not familiar with ZFS and LVM/LUKS on SSDs so please check what I wrote.

Also, I removed the file system note from partitions. I think it still might be worthwhile to mention it in the SSD section.

q66 commented:
It generally looks fine, that said, I wouldn't advise against automatic TRIM like this, for either fstab `discard` or ZFS; on decent SSDs without broken NCQ automatic TRIM is perfectly reasonable and preferred, so this will vary SSD by SSD and cannot be generalized like that.

anjandev commented:
@q66 I updated the patch according to your latest requests. I am hesitant to change the recommendation to prefer periodic trim. Most distros use and recommend periodic trim (ie. debian and ubuntu).

https://wiki.debian.org/SSDOptimization#Mounting_SSD_filesystems

https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/Storage_Administration_Guide/ch02s04.html

q66 commented:
Note how the Debian wiki doesn't say anything about what's preferred or not. You should do like the Debian wiki does, link to the kernel blacklist, and say something along the lines of, "you can use either continuous or periodic TRIM, but usage of continuous TRIM is discouraged if you have an SSD that doesn't handle NCQ correctly, refer to the blacklist" or something.

RHEL documentation is not a good resource, e.g. RHEL 7 is still stuck on kernel 3.10 from 2013.

anjandev commented:
@q66 Thanks for the feedback. That is good reasoning. I have implemented your requested changes.

q66 commented:
Looks fine to me, I'll let others read through it and suggest possible style changes and whatnot, as I'm not the right person to ask for this; perhaps one more thing I can think of, maybe you should put a note in there that `lvm.conf` already contains the required option and you should uncomment change it to the right value, rather than adding a new thing. It's a large config file and it's structured, so people might not see it immediately.

crdpa commented:
Shouldn't have the command to update grub after editing /etc/default/grub ?

ericonr commented on src/config/ssd.md:
```diff
@@ -0,0 +1,150 @@
+If DISC-GRAN (discard granularity) and DISC-MAX (discard max bytes) columns are
+non-zero, that means the block device has TRIM support. If your solid state
+drive partition does not show TRIM support, please verify that you chose a file
+system with TRIM support (ext4, btrfs, etc.).
```
For completion's sake, and because it was created specifically for flash memory, it would be cool to list F2FS filesystem too.

anjandev replied:
Just added a note before running fstrim about having kernel 4.19 or greater as that is the only requirement for using f2fs trim.

ericonr commented on src/config/ssd.md:
````diff
@@ -0,0 +1,153 @@
+We can automate running TRIM by using cron or by adding the `discard` option to
+fstab.
+
+## Periodic TRIM with Cron
+
+Add the following lines to `/etc/cron.daily/fstrim`.
+
+```
+#!/bin/sh
+
+fstrim /
+```
+
+Finally, set the execution bit for the script:
+
+```
+# chmod u+x /etc/cron.daily/fstrim
+```
+
````
Do you have experience with using snooze? I think it would be cool to have more native instructions on how to setup a service like this. If you don't, I think this is okay to merge, mainly because ZFS and Dracut information tend to be a bit sparse.

ericonr replied:
Snooze can use cron jobs, so it isn't a problem. I think you can disregard this.

ericonr commented:
@flexibeast do you think it's ok to merge this one?

flexibeast commented on src/config/ssd.md:
````diff
@@ -0,0 +1,153 @@
+$ lsblk --discard
+```
+
+If DISC-GRAN (discard granularity) and DISC-MAX (discard max bytes) columns are
````
i would add "the" before "DISC-GRAN" here, and write "maximum" in full.

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+
+To run TRIM one-shot, you can run
+[`fstrim(8)`](https://man.voidlinux.org/fstrim.8) manually. For example, if your
+/ directory is on a SSD, run the following command to run TRIM one-shot:
```
Suggest truncating to just "... is on an SSD:" (Note replacement of "a" with "an".)

flexibeast commented on src/config/ssd.md:
````diff
@@ -0,0 +1,153 @@
+```
+
+We can automate running TRIM by using cron or by adding the `discard` option to
+fstab.
````
\`/etc/fstab\`

flexibeast commented on src/config/ssd.md:
````diff
@@ -0,0 +1,153 @@
+# fstrim /
+```
+
+We can automate running TRIM by using cron or by adding the `discard` option to
````
Change voice:

"To automate running TRIM, use cron, or add the ..."

flexibeast replied:
Oh, actually, i see we go into further details about `discard` below. So just change the voice.

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+We can automate running TRIM by using cron or by adding the `discard` option to
+fstab.
+
+## Periodic TRIM with Cron
```
Lowercase "cron".

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+
+## Periodic TRIM with Cron
+
+Add the following lines to `/etc/cron.daily/fstrim`.
```
":" instead of "."

flexibeast commented on src/config/ssd.md:
````diff
@@ -0,0 +1,153 @@
+fstrim /
+```
+
+Finally, set the execution bit for the script:
````
"and make the script executable:"

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+Edit `/etc/fstab` and add the `discard` option to block devices that need TRIM.
+
+For example, if `/dev/sda1` was an SSD partition, formatted as ext4, and mounted
+to `/` with discard options, add the `discard` option:
```
s/to/at/
Remove all text from "with discard" up to, but not including, the colon.

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+
+## LVM
+
+To enable TRIM for LVM's commands (`lvremove`, `lvreduce`, etc) commands, we
```
"etc."
Remove second "commands".

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+## LVM
+
+To enable TRIM for LVM's commands (`lvremove`, `lvreduce`, etc) commands, we
+must open `/etc/lvm/lvm.conf`, uncomment the `issue_discards` option, and set it
```
Remove "we must".

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+> implications](https://wiki.archlinux.org/index.php/Dm-crypt/Specialties#Discard/TRIM_support_for_solid_state_drives_(SSD)).
+
+To open an encrypted LUKS device and allow discards to passthrough, open the
+device with the `--allow-discards` option.
```
":" not "."

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+### Non-root devices
+
+Edit `/etc/crypttab` and set the `discard` option for devices on the SSD. For
+example if you have a LUKS device with the name `externaldrive1`, device
```
"example, if"

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+
+Edit `/etc/crypttab` and set the `discard` option for devices on the SSD. For
+example if you have a LUKS device with the name `externaldrive1`, device
+`/dev/sdb2`, and `none` password, add the `discard` option:
```
"and password \`none\`"

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+
+### Root devices
+
+If your root-device is on LUKS, add `rd.luks.allow-discards` to
```
"root device"

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+### Root devices
+
+If your root-device is on LUKS, add `rd.luks.allow-discards` to
+`CMDLINE_LINUX_DEFAULT`. We can do this for grub by editing `/etc/default/grub`
```
"In the case of GRUB, edit \`/etc/default/grub\`:"

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+## ZFS
+
+Before running `trim` on a ZFS pool, ensure that all devices in the pool support
+by running:
```
"support it:"

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+
+If the pool allows `autotrim` (set `off` by default), you can `trim` the pool
+periodically or automatically. To one-shot `trim` `yourpoolname`, run the
+following command:
```
Remove from ", run" up to the colon.

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,153 @@
+
+### Periodic Trim
+
+Add the following lines to `/etc/cron.daily/ztrim`.
```
":" not "."

flexibeast commented on src/config/ssd.md:
````diff
@@ -0,0 +1,153 @@
+zpool trim yourpoolname
+```
+
+Finally, set the execution bit for the script:
````
"and make the script executable:"

flexibeast commented on src/config/ssd.md:
````diff
@@ -0,0 +1,153 @@
+# chmod u+x /etc/cron.daily/fstrim
+```
+
+## Continuous TRIM with Fstab discard
````
"fstab"

flexibeast commented on src/config/ssd.md:
````diff
@@ -0,0 +1,153 @@
+GRUB_CMDLINE_LINUX_DEFAULT="rd.luks.allow-discards"
+```
+
+### Verifying Configuration
````
"configuration", assuming we aren't using title case for headings (and if we are, other heads will need to be changed to use title case where appropriate).

anjandev replied:
I am only using title case for the first heading (`Solid State Devices`)

flexibeast commented on src/config/ssd.md:
````diff
@@ -0,0 +1,153 @@
+# zpool trim yourpoolname
+```
+
+### Periodic Trim
````
"TRIM"

flexibeast commented:
@ericonr: Review complete. i can review again once the suggested changes have been applied.

anjandev commented:
@flexibeast  requested changes have been made

anjandev commented:
@flexibeast can you please give this a final review?

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,149 @@
+> [security
+> implications](https://wiki.archlinux.org/index.php/Dm-crypt/Specialties#Discard/TRIM_support_for_solid_state_drives_(SSD)).
+
+To open an encrypted LUKS device and allow discards to passthrough, open the
```
"pass through"

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,149 @@
+/ directory is on an SSD:
+
+> Note: To run fstrim on F2FS file systems, you need to be running kernel 4.19
+> or greater.
```
We've recently updated [the style guide with a policy on notes](https://docs.voidlinux.org/contributing/void-docs/style-guide.html#notes), so this text should not use blockquoting.

To not interrupt the flow, i'd like this sentence to immediately follow "manually.", and begin with: "Note that to run \`fstrim\` on ..."

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,149 @@
+> Note: You can use either continuous or periodic TRIM, but usage of continuous
+> TRIM is discouraged if you have an SSD that doesn't handle NCQ correctly.
+> Refer to the
+> [blacklist](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/ata/libata-core.c#n4522).
```
Remove blockquoting.

flexibeast commented on src/config/ssd.md:
```diff
@@ -0,0 +1,149 @@
+
+> Note: Before enabling discard for your LUKS partition, please be aware of the
+> [security
+> implications](https://wiki.archlinux.org/index.php/Dm-crypt/Specialties#Discard/TRIM_support_for_solid_state_drives_(SSD)).
```
Remove blockquoting, and change "Note:" to bolded "Warning:".

flexibeast commented:
@anjandev: We're almost there! Just a few changes, then i'd like @ericonr to check if i've missed anything before merging.

anjandev commented:
Changes have been implemented. Thanks for the review.

ericonr commented:
Looks good to me. I think you still need to implement the changes @flexibeast suggested, right?
